proposals/0029-cooperative-vector.md

```diff
@@ -687,28 +687,52 @@ typedef struct D3D12_COOPERATIVE_VECTOR_PROPERTIES_TRAINING
+{
+    D3D12_COOPERATIVE_VECTOR_PROPERTIES_TYPE_MATRIX_VECTOR_MUL_ADD,
+    D3D12_COOPERATIVE_VECTOR_PROPERTIES_TYPE_OUTER_PRODUCT_ACCUMULATE,
+    D3D12_COOPERATIVE_VECTOR_PROPERTIES_TYPE_VECTOR_ACCUMULATE,
+} D3D12_COOPERATIVE_VECTOR_PROPERTIES_TYPE;
+
 typedef struct D3D12_FEATURE_DATA_COOPERATIVE_VECTOR
-{    
-    InOut UINT                                         MatrixVectorMulAddPropCount;
-    Out D3D12_COOPERATIVE_VECTOR_PROPERTIES_INFERENCE* pMatrixVectorMulAddProperties;
-    InOut UINT                                         OuterProductAccumulatePropCount;
-    Out D3D12_COOPERATIVE_VECTOR_PROPERTIES_TRAINING*  pOuterProductAccumulateProperties;
-    InOut UINT                                         VectorAccumulatePropCount;
-    Out D3D12_COOPERATIVE_VECTOR_PROPERTIES_TRAINING*  pVectorAccumulateProperties;
-};
+{
+    [annotation("_Out_")] BOOL ConfigurationSupported;
+    D3D12_COOPERATIVE_VECTOR_PROPERTIES_TYPE PropertiesType;
+    union
+    {
+        D3D12_COOPERATIVE_VECTOR_PROPERTIES_INFERENCE MatrixVectorMulAddProperties;
+        D3D12_COOPERATIVE_VECTOR_PROPERTIES_TRAINING OuterProductAccumulateProperties;
+        D3D12_COOPERATIVE_VECTOR_PROPERTIES_TRAINING VectorAccumulateProoperties;
+    };
+} D3D12_FEATURE_DATA_COOPERATIVE_VECTOR;
 
 ```
 
 Support for the Cooperative Vector feature is queried through
-`CooperativeVectorTier`. User can also query properties supported for each
-intrinsic in `D3D12_FEATURE_DATA_COOPERATIVE_VECTOR`. If pProperties is NULL
-for any intrinsic, the count of available properties will be returned in
-PropCount. Otherwise, PropCount must represent the size of the pProperties
-array, which will be updated with the number of structures written to
-pProperties upon return. If pProperties is non-NULL for any intrinsic but its
-PropCount is less than the number of properties available for that intrinsic,
-the operation fails and `E_INVALIDARG` is returned.
+`CooperativeVectorTier`. 
+
+The app can also query properties supported for each intrinsic via 
+`D3D12_FEATURE_COOPERATIVE_VECTOR` / `D3D12_FEATURE_DATA_COOPERATIVE_VECTOR`. 
+The `PropertiesType` field selects which corresponding union entry applies 
+for the user to set up a configuration it wants to ask the driver about.
+The driver will set the `ConfigurationSupported` field in the struct on return
+from the all.
+
+`D3D12_COOPERATIVE_VECTOR_PROPERTIES_INFERENCE` has a `BOOL TransposeSupported`
+field - if the app sets this to `TRUE` they are specifically asking for 
+the specified inference configuration with transpose.  The driver support is
+still reported out via the `ConfigurationSupported` field mentioned above 
+(as opposed to treating `TransposeSupported` as an output field).
+
+> To avoid apps having to guess any feature combination that might work,
+> a minimum required set is listed below (though `TransposeSupported` would
+> still need to be checked if needed).  For combinations other than the
```

Comment:
Consider a comment that, based on the cooperative vectors tier, if `TransposeSupported` is false and the configuration is in the minimum supported set, that D3D can/will answer the support question without asking the driver, so the driver only needs to report configurations above and beyond the baseline?

Reply:
Added a note about this, though didn't incorporate tiering into it based on currently defined tiers.  Did I miss something?

Reply:
Yeah that's fine, the tier only impacts the training queries. For tier 0, we can just flat return false for all training queries, and tier 1 adds a minimum baseline there. I wouldn't be surprised to see future tiers that expand out the minimum support set over time.

Reply:
Couple of points about this:
1. If apps have to rely on vendor documentation for this guess and check approach, that is an indirection that is not desirable, also this documentation needs to be updated for different skus and potentially driver versions making it hard to maintain.
2. How will testing work in this case. The current plan is to test beyond the minimum support set, by iterating over the table returned by the driver. Same with caps viewer functionality.

Reply:
Is there a risk the number of supported permutations could explode?

Reply:
Practically I don't think this will happen and the number of combinations will stay in the 10s. We do expect the set to increase as support for newer (smaller) types are more readily available, but unlikely that implementations would advertise support and maintain 100s of combinations. Adding a helper query seems like a good idea.. perhaps down the line?